Node-to-node TLS support

NicolasT opened the pull request:
These changes add (optional) node-to-node TLS encryption & authentication using certificates.

Some TODOs left, but what's here should be OK for review.

CI running without actually _using_ TLS (i.e. no configuration, using plain sockets only), base & quick passed.

TODO:
- [x] Add `mkcerts.sh` script for CI purposes
- [x] Add simple testcase to validate nodes use TLS when configured to, e.g. using `echo -n "" | openssl s_client -connect localhost:4010 -CAfile ca/cacert.pem -cert ca/arakoon1.pem -key ca/arakoon1.key 2>&1 | grep "Verify return code"`
- [ ] Add more elaborate testcases for reconnecting behavior etc (re-use system tests)

NicolasT committed:
Include 'lwt.ssl' dependency

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
CI: Install libssl-dev & OCaml 'ssl' package

NicolasT committed:
Add mkcerts.sh

Note this utility script should not be used to generate certificates
etc. for a real production setup. It should be considered insecure, has
not been audited,...

NicolasT committed:
Add tls_cert, tls_key and tls_ca_cert configuration options

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
Add TLS support to __open_connection

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
Implement node-to-node TLS connectivity

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

domsj commented:
LGTM, but honestly I don't know what exactly I should watch out for ...

NicolasT committed:
Add basic TLS validation test

toolslive commented on src/node/node_main.ml:
```diff
@@ -225,7 +227,30 @@ let only_catchup (type s) (module S : Store.STORE with type t = s) ~name ~cluste
+  match get_ssl_cert_paths me cluster with
+    | None -> None
+    | Some (tls_cert, tls_key, tls_ca_cert) ->
+        let ctx = Typed_ssl.create_both_context Ssl.TLSv1 in
```
Are you sure about TLSv1 ?

NicolasT replied:
``` ocaml
(** Protocol used by SSL. *)
type protocol =
  | SSLv23 (** SSL v3 protocol but can rollback to v2 *)
  | SSLv3 (** SSL v3 protocol *)
  | TLSv1 (** TLS v1 protocol *)
```

Pick your poison...

toolslive commented on src/msg/tcp_messaging.ml:
```diff
@@ -53,7 +53,11 @@ module RR = struct
-class tcp_messaging ?(timeout=60.0) my_addresses my_cookie (drop_it: drop_function) max_buffer_size =
+class tcp_messaging
+  ?(timeout=60.0)
+  ?(client_ssl_context:[> `Client ] Typed_ssl.t option)
```
nice.

NicolasT committed:
Add client service TLS support

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
Accept TLS connections without certificate

Certificate checking should be performed at some other level (e.g.
by using `Ssl.Verify_fail_if_no_peer_cert`).

NicolasT committed:
Implement TLS client support for CLI actions

NicolasT committed:
Doc: Add TLS setup instructions

NicolasT committed:
Tests: Use client TLS connections in test_basic_tls

NicolasT commented:
CI running, ready for merge when success & LGTM'ed.

NicolasT commented:
-base & -quick are OK, others running/queued.

toolslive commented:
good to be merged
